graders/r/Dockerfile

```
@@ -18,74 +19,78 @@
-	r-cran-palmerpenguins \
-	r-cran-png \
-	r-cran-profmem \
-	r-cran-rcpparmadillo \
-	r-cran-rex \
-	r-cran-rsqlite \
-	r-cran-runit \
-	r-cran-shiny \
-	r-cran-stringdist \
-	r-cran-testthat \
-	r-cran-tidyverse \
-	r-cran-tinytest \
-        r-cran-ttdo \
-        r-cran-unix \
-	r-cran-xts \
-	sqlite3 \
-        sudo \
-        && echo "options(diffobj.brightness=\"dark\")" >> /etc/R/Rprofile.site
+# Install required libraries -- using prebuild binaries where available. Because we
+# use r2u, the (R package) name can also be mapped to the (Ubuntu) binary provided
+# by r2u so instead of, say, 'apt install r-cran-tidyverse' we can now say 'install.r
+# tidyverse'. Switching to using the CRAN package names may make it easier to other R
+# users to adapt and adjust the package list.  We also install sqlite3 to support the
+# SQL lectures. (Adding a duckdb binary may be an option too.)
+RUN apt update \
+    && apt install -y curl \
+                      git \
+                      pandoc \
+                      sqlite3 \
+                      sudo \
+    && install.r bench \        
+                 data.table \
+                 devtools \
+                 diffobj \
+                 doparallel \
+                 duckdb \
+                 dygraphs \
+                 flexdashboard \
+                 foreach \
+                 fs \
+                 future.apply \
+                 gapminder \
+                 gh \
+                 git2r \
+                 igraph \
+                 lahman \
+                 lintr \
+                 memoise \
+                 microbenchmark \
+                 nycflights13 \
+                 palmerpenguins \
+                 png \
+                 profmem \
+                 RcppArmadillo \
+                 rex \
+                 rsqlite \
+                 runit \
+                 shiny \
+                 stringdist \
+                 testthat \
+                 tidyverse \
+                 tinytest \
+                 ttdo \
+                 unix \
+                 xts \
+    && echo "options(diffobj.brightness=\"dark\")" >> /etc/R/Rprofile.site
 
 # Install visualTest from Mango
 RUN installGithub.r MangoTheCat/visualTest
 
 # Install plr (helper package for grading R code in PrairieLearn)
 COPY plr /r_autograder/plr
 RUN install2.r --repos=NULL /r_autograder/plr \
-	&& rm -r /r_autograder/plr
+        && rm -r /r_autograder/plr
 
 # Set up user ag
 RUN useradd ag \ 
-	&& mkdir /home/ag \
-	&& chown ag:ag /home/ag \
-	&& echo "[user]" > /home/ag/.gitconfig \
-	&& echo "	name = Autograding User" >> /home/ag/.gitconfig \
-	&& echo "	email = ag@nowhere" >> /home/ag/.gitconfig \
-	&& chown ag:ag /home/ag/.gitconfig
+        && mkdir /home/ag \
+        && chown ag:ag /home/ag \
+        && echo "[user]" > /home/ag/.gitconfig \
+        && echo "       name = Autograding User" >> /home/ag/.gitconfig \
+        && echo "       email = ag@nowhere" >> /home/ag/.gitconfig \
```

Tiny nit since this got caught up in the removal of tabs: indentation isn't actually required here, so let's just drop it.

We're maybe going down a rabbithole here, but: is there a reason this image goes out of its way to set up a git user? When will the grader ever make commits to a repository?

```suggestion
        && echo "name = Autograding User" >> /home/ag/.gitconfig \
        && echo "email = ag@nowhere" >> /home/ag/.gitconfig \
```

---

Because 'Data Science Programming Methods' covers shell, awk, sed, ... we also hit git which we test the same way via 'quasi unit tests' which the grader runs and accumulates.   So we need the ability to run git from the shell and to compare its output and I think I concluded a while back that for that we do need a git config.

---

Huh, interesting. Thanks for the explanation!

---

Sorry hit 'resolved' here by accident. That is your call, of course.

---

Have you considered using `git config -f /home/ag/.gitconfig user.email ag@nowhere` instead?

---

Instead of the redirect?

---

That command will set up the config file in the appropriate format so you don't need to worry about writing directly to the file. This is used elsewhere in PL too:
https://github.com/PrairieLearn/PrairieLearn/blob/613df286061077064dcf44574bd29d4f767c83a6/Dockerfile#L50-L52

This is a nit, of course, feel free to ignore it.

---

I like that approach of writing the config file via calls to `git config`.  In fact, my slides show the same.   Then again, the exiting code should be robust...  

And I could also copy an existing file in.